### DISCUSSION TITLE
Beta-Test: Anbindung von externer Wallbox

### DISCUSSION THREAD

**Discussion by ledermann** [solectrus/solectrus#2865] [Show and tell]
**UPDATE 16.07.2024:** Der Beta-Test ist abgeschlossen, Version 0.15 ist offiziell verfügbar. Eine [aktualisierte Anleitung zur Integration der externen Wallbox](https://github.com/solectrus/hosting/wiki/Integration-einer-externen-Wallbox) in eine bestehende Installation findet sich im Wiki.

Nachfolgend der ursprüngliche Diskussionsverlauf:

----

Analog zu #2861 ist nun auch die Anbindung einer externen Wallbox möglich - also eine Wallbox, deren Verbrauch im Hausverbrauch enthalten ist und die man herausgerechnet haben möchte. Auch das würde ich gerne testen lassen, bevor ich ein neues offizielles Release erstelle. Ping @bennysweb, @jbaesner :-)

Bislang konnte SOLECTRUS nur interne Wallboxen anzeigen, die also auf den Hausverbrauch "oben drauf" kommen, beispielsweise die SENEC-Wallbox. 

Jetzt ist folgendes möglich: Die Verbrauchswerte einer externen Wallbox kommen per MQTT in die InfluxDB von SOLECTRUS. Über eine zusätzliche Umgebungsvariable kann man festlegen, dass der Verbrauch vom Hausverbrauch abgezogen werden soll.

Voraussetzungen:

- Funktionsfähige SOLECTRUS-Installation
- Betrieb einer Wallbox
- Möglichkeit, den Stromverbrauch der Wallbox per MQTT abzufragen
- Grundverständnis von Docker, MQTT, InfluxDB

Hier nun die Anleitung:

## 0. MQTT-Broker vorbereiten

Stelle sicher, dass der Wallbox-Verbrauch per MQTT verfügbar ist. Da das total individuell ist, kann ich hier keine konkrete Anleitung geben. Irgendwas mit ioBroker, evcc & Co.

## 1. MQTT-Collector konfigurieren

In der `.env` muss irgendwo diese Zeile stehen:

```
MQTT_TOPIC_WALLBOX_CHARGE_POWER=somewhere/WALLBOX/POWER
```

Wesentlich ist, das korrekte Topic anzugeben, unter dem der Stromverbrauch der Wallbox abgelegt wird. Die Angabe hier ist nur ein Beispiel. Was genau das einzutragen, ist bei jedem individuell verschieden.

In der `docker-compose.yml` muss unter `services.mqtt-collector.environment` folgende Zeile stehen:

```
- MQTT_TOPIC_WALLBOX_CHARGE_POWER
```

Damit wird dem MQTT-Collector-Container die Variable bekannt gemacht.

Falls du etwas geändert hast, müssen die Container von SOLECTRUS anschließend neu erstellt werden:

```bash
docker compose up -d
```

Nun bitte die Logs des MQTT-Collector-Containers beobachten. Da sollte jetzt auch der Stromverbrauch der Wallbox auftauchen. Wenn das nicht der Fall sein sollte, bitte nicht weiter machen, sondern erst die Ursache klären und beheben.

Der MQTT-Collector schreibt den Wert in das gleiche Measurement der InfluxDB, wie alle anderen Werte. Das wird zukünftig vermutlich konfigurierbar sein, aber derzeitig ist es so. Bitte merke dir, welches Measurement das ist - siehe Variable `INFLUX_MEASUREMENT` bzw. `INFLUX_MEASUREMENT_PV`. Ich gehe im Folgenden davon aus, dass es `my-pv-measurement` heißt. Bei dir könnte es anders sein.

Bevor es weitergeht, solltest du nach Möglichkeit prüfen, ob der neue Messwert wirklich in InfluxDB ankommt (Login im Admin-UI) und im erwarteten Measurement/Field steht.

## 2. SENEC-Collector anpassen

Falls du **auch** den SENEC-Collector benutzt (andernfalls gehe direkt zu Schritt 3), dann muss dieser so eingestellt werden, dass er das Feld `wallbox_charge_power` **nicht** an InfluxDB übermittelt. Andernfalls wird der vom MQTT-Collector gemeldete Messwert immer wieder mit 0 überschrieben.

Hierfür ist ~ein Update des SENEC-Collectors auf die Entwickler-Version~ eine zusätzliche Konfiguration erforderlich. Das geht wie folgt:

<del>
In der `docker-compose.yml` diese Zeile ...

```
image: ghcr.io/solectrus/senec-collector:latest
```

... ändern auf:

```
image: ghcr.io/solectrus/senec-collector:develop
```
</del>

**Ergänzung 29.04.2024**: Die `develop`-Version ist nicht mehr nötig, seit [Version 0.14.0](https://github.com/solectrus/senec-collector/releases/tag/v0.14.0) des SENEC-Collectors kann dieser alles, was nötig ist. Es ist also möglich (und empfehlenswert), den SENEC-Collector wieder auf `latest` einzustellen. Das Folgende bezüglich `SENEC_IGNORE` gilt natürlich weiterhin.

Außerdem muss in der `.env` diese Zeile ergänzt werden:

```
SENEC_IGNORE=wallbox_charge_power
```

Abschließend in der `docker-compose.yml` unter `services.senec-collector.environment` folgende Zeile ergänzen:

```
- SENEC_IGNORE
```

Damit wird dem SENEC-Collector-Container die neue Variable bekannt gemacht.

Jetzt müssen die Container von SOLECTRUS neu erstellt werden:

```bash
docker compose up -d
```

Nun bitte die Logs des SENEC-Collector-Containers beobachten. Da sollte jetzt ein Hinweis erscheinen, dass das Feld ignoriert wird. Wenn das nicht der Fall sein sollte, bitte nicht weiter machen, sondern erst die Ursache klären und beheben.

## 3. SOLECTRUS updaten auf Entwicklungs-Zweig [PR-2836](https://github.com/solectrus/solectrus/pull/2836):

In der docker-compose.yml diese Zeile ...

```
image: ghcr.io/solectrus/solectrus:latest
```

... ändern auf:

<del>

```
image: ghcr.io/solectrus/solectrus:pr-2836
```

</del>

**Update 30.05.2024:** Der PR #2836 wurde mitterweile abgeschlossen und gemerged. Der Code befindet sich also im regulären Entwicklerungszweig `develop`. Das Docker-Image lautet also:

```
image: ghcr.io/solectrus/solectrus:develop
```

In der .env ergänzen:

```
INFLUX_EXCLUDE_FROM_HOUSE_POWER=WALLBOX_POWER
```

Mit dieser **neu eingeführten** Variable wird festgelegt, dass das Dashboard in der InfluxDB den Stromverbrauch der Wallbox vom Hausverbrauch abziehen soll.

Falls du Wärmepumpe **und** externe Wallbox hast und beide Werte abziehen willst, dann geht das auch, und zwar so:

```
INFLUX_EXCLUDE_FROM_HOUSE_POWER=WALLBOX_POWER,HEATPUMP_POWER
```

In der `docker-compose.yml` unter `services.app.environment` folgende Zeile ergänzen:

```
- INFLUX_EXCLUDE_FROM_HOUSE_POWER
```

Damit wird dem App-Container die neue Variable bekannt gemacht.

Anschließend SOLECTRUS neu starten:

```
docker compose up -d
```

Jetzt unbedingt die Logs des App-Containers beobachten. Da findet sich beim Start ein paar Warnungen über fehlende Sensoren, die aber ignoriert werden können. Ansonsten sollte SOLECTRUS wie gewohnt starten.

## 3. Fertig

Das war's schon ;-) Wenn alles richtig gemacht wurde und das ganze MQTT-Gerödel funktioniert, erscheint im SOLECTRUS-Dashboard der Stromverbrauch der Wallbox und der Hausverbrauch reduziert sich entsprechend.

Bin gespannt auf die Rückmeldungen!

**Comment by bennysweb** [comment]
Leider scheitere ich bereits bei der Installation des MQTT Collectors. Wie bekomme ich ihn als Docker-Instanz auf der Synology installiert? Kann ich die Daten einfach in das bereits bestehende Solectrus docker-compose File einfügen und die bereits bestehende Solectrus .env ergänzen?

**Reply by ledermann** [reply]
Ja, genau so. Der MQTT-Collector muss zunächst in die `docker-compose.yml` integriert werden, Beispiel hier:
https://github.com/solectrus/hosting/blob/main/guide/mqtt-evcc/docker-compose.yml#L118-L158

Die bestehende `.env` wird dann ebenfalls erweitert, Beispiel hier:
https://github.com/solectrus/hosting/blob/main/guide/mqtt-evcc/.env#L73-L93

**Comment by bennysweb** [comment]
Aaah, ja, danke. Das hat schonmal geholfen. Der Collector läuft. Kann es sein, dass das Topic „MQTT_TOPIC_WALLBOX_CHARGE_POWER“ heißen muss, und nicht „MQTT_TOPIC_WALLBOX_POWER“?

**Reply by ledermann** [reply]
Ja, du hast recht. Ich bin für das nächste SOLECTRUS-Update dabei, die Bezeichner zu vereinheitlichen. Ich habe oben die neuen Namen aufgeführt, die Collectors verwenden aber noch die alten. Sorry. Im MQTT-Collector muss es also `MQTT_TOPIC_WALLBOX_CHARGE_POWER` heißen. Ich ändere das in der obigen Anleitung ab.

Ansonsten noch: Da du die meisten Werte für SOLECTRUS über den SENEC-Collector bekommst, wird der MQTT-Collector nur für die Wallbox benötigt (und ggfs. Wärmepumpe, falls vorhanden). Somit solltest du darauf achten, nur die benötigten Topics aufzuführen, also vermutlich nur `MQTT_TOPIC_WALLBOX_CHARGE_POWER`.

**Comment by bennysweb** [comment]
Okay, die Werte von der openWB werden erfolgreich empfangen und landen auch in der Datenbank. Super :-)
Weitere Frage: Muss man eventuell dem Senec-Collector sagen, dass er keine „0“-Werte mehr in die Datenbank schreiben soll? 

￼
Die Werte mit 4000 stammen von der Wallbox…

Das E-Auto wird in der App auch nur kurz angezeigt und verschwindet wieder, taucht wieder auf, usw...

**Reply by ledermann** [reply]
Wie gut, dass wir diese Testphase machen - du hast offenbar einen Use-Case, den ich bislang nicht bedacht habe: Der SENEC-Collector schreibt permanent eine 0, weil es bei dir keine SENEC-Wallbox gibt. Der MQTT-Collector schreibt parallel den richtigen Wert der externen Wallbox ins gleiche Feld der InfluxDB. 

Das ist suboptimal. Ich stimme dir zu, der SENEC-Collector muss aufhören, Nullen zu schreiben. Ich überlege mir dazu etwas. Vielleicht wird es eine neue Option in der Konfiguration geben müssen.

Übrigens ist dein Screenshot hier nicht zu sehen. Besser direkt in GitHub schreiben, das Antworten auf die Mails ist nicht so ideal.

**Reply by bennysweb** [reply]
Okay, das ist super, danke 😀
Und danke auch für den Hinweis mit dem Screenshot, werde ich nächstes Mal beachten.

**Reply by bennysweb** [reply]
Mir ist noch etwas aufgefallen: Auch wenn ich keine Wärmepumpe konfiguriert habe, gibt es in der Tages-, Wochen-, Jahres- und Gesamtübersicht eine ganz kleine Anzeige auf der Verwendungsseite. 
Hier ein ordentlicher Screenshot ;-) :
<img width="1284" alt="Bildschirmfoto 2024-02-29 um 21 45 44" src="https://github.com/solectrus/solectrus/assets/77127461/ee4aa1c8-27b4-4162-9fe2-b12274d6b51f">

**Reply by ledermann** [reply]
So, der SENEC-Collector kann nun so [eingestellt](https://github.com/solectrus/senec-collector/blob/develop/.env.example#L28-L31) werden, dass einzelne Messwerte **ignoriert** werden. Die Anleitung oben habe ich um einen weiteren Schritt ergänzt.

Den kleinen Darstellungsfehler im Dashboard habe ich auch korrigiert - Messwerte für die Wärmepumpe werden in der Strombilanz nur angezeigt, wenn eine Wärmepumpe auch konfiguriert ist.

@bennysweb Klappt es damit?

**Reply by bennysweb** [reply]
Ja, das funktioniert! Wow 😄 Muss ich für die Änderungen am Dashboard nochmal auf eine andere Version wechseln? Jetzt muss ich bloß noch aus influx die fehlerhaften 0-Werte von gestern löschen, dann passt das. Top 👍 
Edit: Ein docker-compose pull hat geholfen. Jetzt schaut es wie erwartet aus.

**Reply by ledermann** [reply]
Prima :) Nicht andere Version, nur Update durchführen: `docker compose pull && docker compose up -d`. Grundsätzlich empfehle ich WatchTower, damit man das nicht immer manuell machen muss. 

Löschen aus InfluxDB ist allerdings keine einfache Sache bzw. nur durch Überschreiben an den exakt gleichen Zeitstempeln möglich. Dafür habe ich leider keine Lösung anzubieten.

**Reply by bennysweb** [reply]
Aah, du warst schneller. Jetzt passt es. 👍

**Reply by bennysweb** [reply]
Und dein nächster Kaffee geht auf mich ☕️

**Reply by bennysweb** [reply]
Noch etwas, was mir eingefallen ist: Damit die Sache rund wird, müsste auch der csv-importer angepasst werden. Sonst ergänzt der in der Datenbank wieder die "böse" 0.

**Reply by ledermann** [reply]
Danke sehr :) Ja, der CSV-Importer braucht eine Anpassung, ich habe https://github.com/solectrus/csv-importer/issues/171 angelegt.

**Comment by bennysweb** [comment]
**Feature Request: SoC E-Auto**
Als optional mögliche Anzeige in Solectrus wäre es toll, wenn auch der Ladestand des E-Autos mit angezeigt wird. Ich weiß nicht, ob Senec diesen bereitstellt, aber über den MQTT-Collector könnte das möglich sein. Die Anzeige könnte ich mir z.B. mittig unter der Senec-Akku-Anzeige sehr gut vorstellen.

**Reply by ledermann** [reply]
Denkbar, ja, aber das muss für jede Automarke oder gar jedes Automodell separat gelöst werden, sofern diese eine Schnittstelle dafür bereitstellen. Die Wallbox oder der Speicher weiß nichts über den Akku des Autos. Für einen Renault Zoe habe ich das gemacht, der Wert wird in SOLECTRUS aber bislang nicht angezeigt.

Sinnvoll wäre vielleicht, die Messwerte von woanders (z.B. evcc) per MQTT nach SOLECTRUS zu bringen. 

Ich habe mal https://github.com/solectrus/solectrus/issues/2881 angelegt, ist aber low-prio.

**Comment by Deanw1975** [comment]
Hallo Georg,
interessante Entwicklung!

Werde ich gerne testen.

Würde aber noch Impulse geben wollen.

Sorry für die nun folgend lange Ausführung, denke es ist wichtig um meine Gedanken darzulegen.

****Impuls 1**: Strom ging an welche Wallbox**
Es wäre Klasse, wenn ich in Solectrus die Wallboxen erkennen kann, die Strom gezogen haben.

Sicherlich haben einige mehrere Wallboxen und würden evtl. Hier sehen wollen, wieviel ging an alle Autos und wieviel ging an ein bestimmtes Auto.

****Impuls 2**: Sonne geladen oder zugekauft**
Dieses Thema würde über Impuls 1 wahrscheinlich in meinem Fall mit dem ersten Impuls abgedeckt.
Wieviel Strom wurde pur aus der Sonne in die Verbraucher gesteckt.
Wie gehe ich mit dem zweiten Impuls um:

Mit folgender Herausforderung stehe ich sicher nicht alleine da.
Installation bei mir:
Senec V3 mit Senec Wallbox

Da mir die Wallboxsteuerung „Solar optimiert“ von Senec nicht zusagt, habe ich EVCC am laufen.
Sonne lasse ich pur über EVCC laden!
Wenn der Wagen schnell voll werden soll, dann lasse ich die Wallbox von Senec steuern. Ich weiß auch EVCC könnte den Wagen ohne Sonne laden, aber dann wären die Daten für mich in Solectrus vermischt und ich muss meine Ladendaten in EVCC manuell auseinander nehmen. Deswegen habe ich Solectrus.

Solecturs lief bis vor ca. 1 Woche nur auf einem PI mit der Anbindung via MQTT Anbindung (EVCC). Somit habe ich die Ladedaten (Sonne laden) in Solectrus sehen können. Die Ladedaten ohne Sonne in der Senec App. 
In Excel habe ich dann aus der CSV von Senec eingelesen und die Daten aus EVCC CSV eingelesen. Somit wusste ich wieviel Sonne ich geladen habe und wieviel ich fürs Auto zugekauft habe. 
Jeder Zeit konnte ich das auch via Solectus und Senec APP überschlagen.

Seit ca. einer Woche teste ich Proxmox und lasse dort Solctrus in einem Container mit der Senec Anbindung laufen.
Umzug des PI (EVCC Anbindung) auf Proxmox steht noch aus. 

Nun habe ich also die Daten in ZWEI Solectrus Installationen, kann also jederzeit auf einem Blick die gewünschten Infos einholen.

Die Frage aber: wieviel ging vollstäding in den Wagen kann ich nicht auf einem Blick festellen.
Mit deinem Ansatz kann ich das umsetzen! > **DANKE**

Hierzu muss ich aber nun eine dritte Installation heranziehen. Unter Proxmox kein Thema.

Deswegen meine Impulse.
Für Solectrus wäre mein Umgang zwei "Walboxen" die unterschiedlich Ihren Strom erhalten.
Ein klick auf Wb1 (Senec Steuerung)= verbrauch via Netz / Ein Klick Wb2 (EVCC Steuerung) = verbrauch Sonne.
Ein Klick auf "alle WB" Stromverbrauch gesamt

Viele Grüße
Dean

**Reply by ledermann** [reply]
Danke für deine Ausführungen, ich schätze das sehr! Dass du mit 2-3 SOLECTRUS-Instanzen jonglierst, ist beeindruckend :-)

Die Unterstützung mehrerer Wallboxen ist geplant. Ich möchte aber erst mal die jetzige Implementierung (externe Wallbox sowie Wärmepumpe) rund bekommen. Der wesentliche Aspekt ist das Herausrechnen aus dem Hausverbrauch.

Zur Berechnung des "Sonne-Ladens" passt in etwa #2436. Ja, das finde ich auch sehr spannend: Welcher Anteil des Verbrauchs kommt aus dem Netz und welcher von der Sonne? Die Ideen gehen nicht aus :)

**Comment by Deanw1975** [comment]
Kleine Ergänzung noch:
Evcc kennt auch die Kosten die für die Ladung entstehen und überträgt diese via MQTT. Diese könnten evtl. von Solectrus beachtet werden. Das wäre ein Ansatz der dieses Thema zumindest in teilen lösen könnte:
https://github.com/orgs/solectrus/discussions/2854

LG
Dean

**Comment by q2doc** [comment]
@ledermann, zwei Wallboxen parallel gehen damit noch nicht korrekt?

**Reply by ledermann** [reply]
Momentan ja. Derzeitig kennt SOLECTRUS nur **eine** Wallbox. Langfristig möchte ich aber auch mehrere ermöglichen.

**Comment by k-ranseier** [comment]
Halle allerseits.
Erstmal vielen Dank für das tolle Programm!

Da ja Solectrus und der Senec-collector mittlerweile aktualisiert worden sind frage ich mich ob die Installation der .develop bzw. die .pr-2836  Version noch nötig ist.

Gruss
Karsten

**Reply by ledermann** [reply]
Danke für den Hinweis! Ja, der SENEC-Collector kann seit [Version 0.14.0](https://github.com/solectrus/senec-collector/releases/tag/v0.14.0) alles, was nötig ist. Hier braucht es nicht mehr die `develop`-Version. Ich habe die Anleitung oben angepasst.

Für SOLECTRUS selbst, also die Dashboard-App, ist aber weiterhin die Beta-Version (Tag `pr-2836`) nötig. Da sich durch die neue Sensor-Konfiguration vieles ändert, möchte ich das noch etwas weiter testen. Insbesondere möchte ich sicherstellen, dass SOLECTRUS auch ohne jeden manuellen Eingriff mit der alten Konfiguration weiter läuft. Sobald das geklärt ist, wird es das Major-Release 0.15 geben und `pr-2836` ist dann obsolet.

**Comment by ledermann** [comment]
Dank der vielen Rückmeldungen denke ich, dass die Sache ausreichend stabil läuft. Ich habe somit den PR #2836 gerade nach `develop` gemerged, das heißt:

**Der Tag `pr-2836` ist jetzt obsolet. Wer diesen nutzt (also alle Tester), bitte auf `develop` wechseln. Das ist gefahrlos möglich.**

Das nächste offizielle größere Update (Version 0.15) befindet sich jetzt auf der Zielgeraden ;-)

**Reply by q2doc** [reply]
Danke!!! Was ist denn aktuell der richtige Tag für den mqtt-collector? Ich bin da noch auf pr-137 ... ah und ich sehe beim Dashboard bin ich noch auf solectrus:next ... einfach lassen und auf die 15 warten?

**Reply by ledermann** [reply]
[PR-137 beim MQTT-Collector](https://github.com/solectrus/mqtt-collector/pull/137) ist auch gemerged. Aktuell ist das hier meine Empfehlung:

- `ghcr.io/solectrus/mqtt-collector:develop`
- `ghcr.io/solectrus/solectrus:develop`

Das `next`-Tag bringt momentan keinen Mehrwert. Das werde ich aber bei zukünftigen Weiterentwicklungen wieder aktivieren.

**Comment by ledermann** [comment]
Version 0.15 mit offizieller Unterstützung der externen Wallbox ist nun da 🎉 

Herzlichen Dank an alle, die mit ihrem Feedback eine so große Unterstützung waren! Wer neu hinzukommt, findet im Wiki eine [aktualisierte Anleitung zur Integration der externen Wallbox in eine bestehende Installation](https://github.com/solectrus/hosting/wiki/Integration-einer-externen-Wallbox).

**Comment by boecks** [comment]
Hallo Georg, ich finde es toll was Solectrus bis jetzt bietet und auch die Aussicht darauf, was es noch können wird. Danke für deine Arbeit und Mühe!
Mich würde in diesem Sinne interessieren, wie es um "beliebige" Großverbraucher und deren Anzeige steht? (bin dem ursprünglichen Issue bis hierher gefolgt 😃)

Ich habe schon Zähler für meinen Heizstab und die Klimaanlage laufen. (Ein Shelly Pro EM 50) 
Zähler für WP und Wallbox werde ich noch installieren. (Werden wohl Shelly Pro 3EM - gut und günstig)

Der MQTT Collector und das Sensor Konzept stehen ja und würden das schon hergeben denke ich. Auch der Powersplitter ist ein größerer Brocken zu dem Thema, der schon steht. 

Was mir bereits auffällt an der seperaten Darstellung (vor allem in deiner Live Demo), dass manche Segmente des rechten Balken so schmal sind, dass man sie kaum wahrnehmen und bedienen kann. Wäre es eine Idee seperat gemessene Verbraucher nur im Tooltipp des Hausverbrauch darzustellen oder zu listen? An der aktuellen Möglichkeit der seperaten Darstellung würde ich aber ebenfalls nichts ändern...

**Reply by ledermann** [reply]
Danke für das Feedback :) Ich habe mal #3396 dafür angelegt.
